### PR TITLE
[megatron] fix: dynamic context parallel batch splitting and loss normalization

### PR DESCRIPTION
**Pull request by Kite0011**
### What does this PR do?

  修复 Dynamic Context Parallel (DCP) 在 Megatron 引擎中的 batch 切分逻辑和 loss 归一化问题。

  相关改动：
  1. 重写 dynamic_cp_split_batch：修复当序列数少于 dp_size 时直接 return 导致 CP
  未生效的问题；改用正确的 power-of-2 CP size 计算逻辑，支持不均匀序列分配
  2. 修复 forward_backward_batch 中 DCP 的 loss 归一化：DCP 模式下 sp_size 应为整个 DP world size，且
  batch_num_tokens 不应在 DP group 内 all-reduce（因为 CP 子组内的 token 是重复的）
  3. Megatron recomputation patch 兼容性：增加 _recover_function_args 检查，避免在旧版 Megatron（无
  saved_objects 支持）上 crash

+ 参数设置：
  + 当开启dynamic_context_parallel时，cp_size需要设置成1
  + 当开启dynamic_context_parallel时，建议max_seqlen_per_dp_cp_rank设置跟max_token_len_per_gpu一样

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

  verl/utils/megatron_utils.py — dynamic_cp_split_batch：
  - local_cp_size 计算：ceil(max_seq_len / max_seqlen_per_dp_cp_rank) 后 round up to power-of-2
  - 新增：当序列数不足以覆盖所有 DP 子组时，自动增大 local_cp_size 保证每个子组至少分到一条序列
  - 数据切分：按 local_dp_rank 均匀分配序列（处理余数），替代原来的 ceil 分配（可能越界）
  - 写入 dp_size=local_dp_size 用于下游 loss 归一化

  verl/workers/engine/megatron/transformer_impl.py — forward_backward_batch：
  - DCP 模式下 sp_size 设为 mpu.get_data_parallel_world_size()（而非 context_parallel_size）
  - DCP 模式下跳过 batch_num_tokens 的 DP all-reduce
  - dp_size 统一使用 mpu.get_data_parallel_world_size()

  verl/models/mcore/patch.py — apply_patch_megatron_recomputation_backward：
  - 增加 hasattr(rd.CheckpointFunction, '_recover_function_args') 前置检查，兼容旧版 Megatron

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

**Comment by Kite0011** [comment]
@ISEEKYAN

File: verl/workers/engine/megatron/transformer_impl.py [hunk 1]
```diff
@@ -581,15 +581,21 @@ def load_checkpoint(
...
         batch_num_tokens = data["loss_mask"].sum().to(get_device_id())
-        torch.distributed.all_reduce(
-            batch_num_tokens, op=torch.distributed.ReduceOp.SUM, group=self.get_data_parallel_group()
-        )
+        if self.engine_config.dynamic_context_parallel:
+            pass
+        else:
+            torch.distributed.all_reduce(
+                batch_num_tokens, op=torch.distributed.ReduceOp.SUM, group=self.get_data_parallel_group()
+            )
```
[the remark this answers is not in the document]
Comment (Kite0011) [reply verl/workers/engine/megatron/transformer_impl.py:596]: 在dcp的场景下，每个卡初始就会拿到所有数据，所以可以跳过这里的allreduce

[the remark this answers is not in the document]
Comment (Kite0011) [reply verl/workers/engine/megatron/transformer_impl.py:596]: Key design: In DCP mode, context_parallel_size is set to 1 in the Megatron config. This means Megatron treats all
  ranks as a single DP group. All ranks receive the identical batch from the dataloader. The dynamic CP grouping happens
  at runtime inside dynamic_cp_split_batch.

  ▎ Concern 1 (skip all-reduce for batch_num_tokens):
  ▎ Since all ranks start with the same batch, loss_mask.sum() already equals the global token count T_total on every
  rank. An all-reduce would incorrectly produce W × T_total. Skipping it is correct.

File: verl/workers/engine/megatron/transformer_impl.py [hunk 2]
```diff
@@ -581,15 +581,21 @@ def load_checkpoint(
             offload_megatron_optimizer(self.optimizer)
 
     def forward_backward_batch(self, data: TensorDict, loss_function: Callable, forward_only=False) -> Any:
-        tu.assign_non_tensor(data, sp_size=self.engine_config.context_parallel_size)
+        if self.engine_config.dynamic_context_parallel:
+            tu.assign_non_tensor(data, sp_size=mpu.get_data_parallel_world_size())
+        else:
+            tu.assign_non_tensor(data, sp_size=self.engine_config.context_parallel_size)
```
[the remark this answers is not in the document]
Comment (Kite0011) [reply verl/workers/engine/megatron/transformer_impl.py:587]: sp_size is not used in the loss formula (loss = -masked_sum / batch_num_tokens * dp_size). It's only consumed in
  prepare_micro_batches() for computing max_token_len = max_token_len_per_gpu * sp_size, which determines micro-batch
  splitting granularity. Setting it to W correctly represents the total token capacity across all ranks.

File: verl/utils/megatron_utils.py [hunk 3]
```diff
@@ -1368,47 +1368,69 @@ def get_megatron_module_device(models: list[Any]) -> str:
...
-            num_seq_per_local_cp = math.ceil(len(seq_len_effective) / local_dp_size)
-            start_idx = local_dp_rank * num_seq_per_local_cp
-            end_idx = min(start_idx + num_seq_per_local_cp, len(seq_len_effective))
-            selected_indices = indices[start_idx:end_idx]
-            batch = tu.index_select_tensor_dict(batch, selected_indices)
-
-    # print(f"rank={torch.distributed.get_rank()}, local_cp_size={local_cp_size} max_seq_len={max_seq_len}")
+    # --- attach metadata used by downstream model forward and loss ---
     tu.assign_non_tensor_data(batch, "local_cp_size", local_cp_size)
+    tu.assign_non_tensor(batch, dp_size=local_dp_size)
```
[the remark this answers is not in the document]
Comment (Kite0011) [reply verl/utils/megatron_utils.py:1430]: This works together with the unchanged batch_num_tokens = T_total. Within a CP subgroup of size C, all ranks compute
  the same loss (outputs are all-gathered via postprocess_thd_engine). After DDP averaging across W = C × D ranks:

  ▎ grad_avg = Σ_j(C × loss_j × D / T_total) / (C × D) = Σ_j(loss_j) / T_total ✓

  ▎ The three changes form a consistent system — batch_num_tokens = T_total (no all-reduce needed), dp_size = D 
  (compensates for C-fold gradient duplication in DDP), and the math produces the correct global average loss.

File: verl/workers/engine/megatron/transformer_impl.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ISEEKYAN) [reply verl/workers/engine/megatron/transformer_impl.py:587]: 这里是不是因为开了动态CP的时候要保证context_parallel_size=1？如果是这种限制的话应该放在config的校验上面

File: verl/workers/engine/megatron/transformer_impl.py [hunk 4]
```diff
@@ -581,15 +581,21 @@ def load_checkpoint(
...
-        )
+        if self.engine_config.dynamic_context_parallel:
+            pass
+        else:
+            torch.distributed.all_reduce(
+                batch_num_tokens, op=torch.distributed.ReduceOp.SUM, group=self.get_data_parallel_group()
+            )
         tu.assign_non_tensor(data, batch_num_tokens=batch_num_tokens.item())
-        tu.assign_non_tensor(data, dp_size=self.get_data_parallel_size())
+        tu.assign_non_tensor(data, dp_size=mpu.get_data_parallel_world_size())
```
Comment (ISEEKYAN) [inline verl/workers/engine/megatron/transformer_impl.py:598]: 这样就不满足逻辑DP=1了，数据会被切分掉吗

Comment (Kite0011) [reply verl/workers/engine/megatron/transformer_impl.py:598]: 简而言之是，这个td里面的dp_size是用来做loss norm的，所以需要拿全局的world_size；数据切分不会用到这个数

  ▎ DCP 模式下，Megatron 看到的是 context_parallel_size=1，所以 DP = W / TP / PP，数据会被按这个
  DP size 切分到各个 rank。之后在 forward 阶段，动态 CP 会把同一个 DP group 内的若干 rank 组成 CP
  subgroup 来协作处理长序列。

  ▎ 这里设 dp_size = local_dp_size（即 W / TP / PP / CP_runtime）是为了 loss
  normalization，不是控制数据切分。因为 CP subgroup 内的 rank 处理的是同一批 token 的不同片段，DDP
   做 gradient all-reduce 时会把这些重复计算的梯度求和，所以需要除以 dp_size = D/C 来补偿 C
  倍的梯度重复，最终得到正确的全局平均 loss。

  ▎ 数学上：loss = sum(loss_i) / (batch_num_tokens × dp_size)，其中 batch_num_tokens =
  T_total（不做 all-reduce，因为每个 rank 已经持有全量 loss_mask），dp_size = D，最终 = sum /
  (T_total × D) 即全局平均。

File: verl/workers/engine/megatron/transformer_impl.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kite0011) [reply verl/workers/engine/megatron/transformer_impl.py:587]: 对的，我在engine.py里加了个assert

**Review by ISEEKYAN — DISMISSED** [review]

**Comment by ETOgaosion** [comment]
@Kite0011 please help run with pre-commit, or run `pre-commit install` to produce syntax correction automatically~

**Comment by Kite0011** [comment]
@ETOgaosion Done~

**Comment by Kite0011** [comment]
fix test lint @ETOgaosion

**Comment by Kite0011** [comment]
@ISEEKYAN Please re-review this pr, I fixed the code formatting

**Comment by Kite0011** [comment]
@ISEEKYAN Please re-re-review this pr, I added code license
